design/proposal-concurrent-collectors.md

```diff
@@ -0,0 +1,55 @@
+# Run collectors concurrently
+ 
+## Goals
+
+Increase the speed at which collectors complete.
+
+Ensure that the cluster-resources collector runs prior to any others, so that things like pod status and pod logs only include the pods from the application rather than the ones that Troubleshoot started as part of collectors.
+## Non Goals
+
+
+## Background
+
+Currently all collectors run sequentially, in the order in which they are specified in the spec.  This is implied, and has not been guaranteed.  With the addition of the ability to supply multiple specs, the order is no longer predicatble.
+
+Although sequence is not guaranteed for any collectors, #768 ensures that the `clusterResources` collector runs first, so that the list of resources does not include those that are started during the Troubleshoot collection sequence.
+
+## High-Level Design
+
+Switch all collectors to use goroutines to collect (concurrently).  This should run the list of collectors of `clusterResources` type first, then the remainder concurrently when `clusterResources` is complete.
+
+## Detailed Design
+
+Tasks that can be split to separate merges (in order):
+
+* Switch all collectors to use goroutines to collect (concurrently).
+
+In `preflight.Collect()` and `supportbundle.runCollectors()`, refactor to use a `go func()` pattern to run all the collectors that are not `clusterResources` concurrently.  Check that the first in the list is `clusterResources` and run that separately, prior to the others.
+
+TODO: review host and remote collectors
+
+## Limitations
+
+Relies on splitting up the list of collectors into `clusterResources` and others.
+
+## Assumptions
+
+* The Kubernetes API can handle the load of collectors running concurrently, including when the pod logs collector runs
```

Comment:
Wonder if for this reason this should be optional in some way? You could have a property in the spec to enable it, but that wouldn't work well with how we combine specs today.

Reply:
I think this is the type of thing we need make a reasonable decision on and not try to pass through to the user to figure out. Appropriate throttling should be something we can make a reasonable decision on and it's unlikely to be something and end user has sufficient information to make any informed decision on.

Reply:
does the k8s API not throttle incoming requests properly?  If that's an issue we would probably want to consider how we resolve it for other instances of high query load.

Maybe we have a concurrency limit that is a reasonable default, optionally some config for it?